### DISCUSSION TITLE
LocalStorage and useState in _app.js for fetched API data

### DISCUSSION THREAD

**Discussion by StefanieLoeser** [neuefische/questions#191] [Web]
Hi everybody, 

after the successful implementation of useState and localStorage for my static dummy-data that was solved in this [question-thread ](https://github.com/neuefische/questions/discussions/176) I now have a follow-up problem after connecting to my collection-api. 

As the fetch can't happen within the _app.js I thought to simply update the state with the fetched data and have an empty array as the initial state for the localStorage functionality. Unfortunately now I get an error occurring within the useLocalStorage-component. 

Thank you!

[Current PR](https://github.com/StefanieLoeser/capstone-project/pull/14)

Error message: 
![Bildschirmfoto 2022-09-29 um 16 05 49](https://user-images.githubusercontent.com/104068626/193053567-ca2cf550-663c-4a9e-8a3d-98d5577e03fb.png)

index.js
![Bildschirmfoto 2022-09-29 um 16 04 19](https://user-images.githubusercontent.com/104068626/193053167-93006563-c86d-43b3-afdc-728ac874163b.png)

app.js
![Bildschirmfoto 2022-09-29 um 16 04 52](https://user-images.githubusercontent.com/104068626/193053318-85c187b2-8568-4c03-b00d-717c28e2183c.png)

useLocalStorage.js
![Bildschirmfoto 2022-09-29 um 16 05 14](https://user-images.githubusercontent.com/104068626/193053429-1bfff357-a5b9-4262-8987-936230adbd95.png)

**Comment by ThomasKirst** [comment]
Hi Stefanie,
could you try to wrap your call to `onSetCollectionState(...)` (`index.js` line 50) within a `useEffect()`?
```js
useEffect(() => onSetCollectionState(myDiscogsCollection), []); // empty dependency array => runs after initial render
```
This code will then be called only on the client side and not during the server render phase (where it cannot access the browser´s window object).

**Reply by StefanieLoeser** [reply]
@ThomasKirst It seemed to work but only for a first render, when I refresh the page it tells me it can't find the properties that I am using for the mapping. 

![Bildschirmfoto 2022-09-29 um 16 55 36](https://user-images.githubusercontent.com/104068626/193066016-0bab4c31-074d-4f82-9f82-2cea4cdeab52.png)

**Reply by NikRoe** [reply]
we solved the issue by making `record.basic_information` optional with optional chaining.

In the relevant places it now looks like this: 
```js
<Cover src={record.basic_information?.cover_image} />
```

**Comment by StefanieLoeser** [comment]
Hello @NikRoe, me again!

Thank you for your help earlier. Unfortunately there came up another follow up issue. Due to extending the fetched data with the "isChecked: false" property the bookmark-localStorage-logic doesn't work correctly anymore. The data gets fetched again when I switch between pages (from collection to recordbag and back) and this rerender then overwrites the made or adjusted selection. 

I tried to fix it by adding the isChecked value within the useEffect and not in the fetch itself but now I get an error message from localStorage. 

The amended code looks like this:
![Bildschirmfoto 2022-09-30 um 11 54 24](https://user-images.githubusercontent.com/104068626/193246509-9a426312-889c-446a-a901-5edea1604691.png)

Error message:
![Bildschirmfoto 2022-09-30 um 12 03 44](https://user-images.githubusercontent.com/104068626/193246671-c73fbb63-8b36-4bd8-b83b-4401e2b5e677.png)
